docs/dev/specs/packable.md

```diff
@@ -0,0 +1,315 @@
+# Motivation
+
+As part of the transition to Coordicide it has been decided that the
+serialization API used by the Bee ecosystem needs to be redesigned. The main
+motivation for this redesign is being able to make most of the crates in Bee
+usable in a `no_std` environment.
+
+## The old `Packable` trait
+
+The need for a serialization API existed before Coordicide. Efforts to satisfy
+this need culminated with the introduction of the `Packable` trait in the
+`bee-common` crate during the Chrysalis part 2 period. Most of the design
+decisions behind this trait were done to simplify the serialization of the
+[IOTA protocol messages](https://github.com/iotaledger/protocol-rfcs/pull/0017).
+The proposed trait was the following
+
+```rust
+pub trait Packable {
+    type Error: Debug;
+
+    fn packed_len(&self) -> usize;
+
+    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error>;
+
+    fn unpack_inner<R: Read + ?Sized, const CHECK: bool>(reader: &mut R) -> Result<Self, Self::Error>
+    where
+        Self: Sized;
+}
+```
+
+The main issue with this trait is that it cannot be used in a `no_std`
+environment because it depends explicitly on the `std::io` API, whose
+transition to the `core` crate has not been decided yet.
+
+Another issue is that the `Error` type is used to represent three different
+kinds of errors:
+
+- Serialization: Raised when there are issues while writing bytes.
+- Deserialization errors: Raised when there are issues while reading bytes.
+- Semantic errors: Raised when the bytes being used to create a value are
+  invalid for the data layout of such value.
+
+# Design
+
+## Replacing `std::io`
+
+We introduce two new traits to replace `Read` and `Write` from `std::io`.
+
+First we have the `Unpacker` trait which represents a type that can be used to
+read bytes from it.
+
+```rust
+pub trait Unpacker: Sized {
+    type Error;
+
+    fn unpack_bytes(&mut self, slice: &mut [u8]) -> Result<(), Self::Error>;
+}
+```
+
+- The `Error` associated type represents any error related to byte
+  deserialization.
+
+- The `unpack_bytes` method writes enough bytes to fill `slice`. This method
+  must fail if the unpacker does not have enough bytes to fulfill the request.
+
+We also have the `Packer` trait which represents a type that can be used to
+write bytes into it.
+
+```rust
+pub trait Packer {
+    type Error;
+
+    fn pack_bytes(&mut self, bytes: &[u8]) -> Result<(), Self::Error>;
+}
+```
+
+- The `Error` associated type represents any error related to byte
+  serialization.
+
+- The `pack_bytes` method reads all the bytes from `slice`. This method must
```

Comment:
```suggestion
- The `pack_bytes` method reads all the bytes from `bytes`. This method must
```

Reply:
Also, I would do the sentence the other way around to use `writes` instead of `reads`, would better match the description of Packer.